### PR TITLE
perf(flydsl): fuse Kimi-K3 FP8 pre-route projections

### PR DESCRIPTION
**Pull request by JohnQinAMD**
## Purpose

Add three fixed-shape gfx950 FlyDSL primitives for Kimi-K3 TP8 batch-one MoE
pre-route: FP8 routed/shared gate-up dual projection, fused SiTU + FP8
shared-down, and a mixed-precision tri-projection returning FP32 router logits.
Activations/outputs remain BF16; row-scaled OCP E4M3 weights and FP32
accumulation are used. Unsupported contracts fail closed.

## Test plan

- Validate all support predicates, dequantized dual/tri references,
  shared-down, graph replay, and exact router top-16 membership.
- Benchmark both the dual/shared and complete tri boundaries against exact
  BF16 parent ROCm/aiter#4498.
- Enforce dual routed/shared RRMSE `<0.035/<0.06`, cosine `>0.999/>0.998`,
  and tri routed/shared/router RRMSE `<0.035/<0.035/<0.01`.

```bash
IMAGE='vllm/vllm-openai-rocm:kimi-k3@sha256:5aa7e626ff73672f5ca7aae46754570488c23d33ca1ac90756a1d2d1a3fe099b'
docker run --rm --ipc=host --shm-size=16g \
  --device=/dev/kfd --device=/dev/dri --group-add=video --group-add=render \
  -e HIP_VISIBLE_DEVICES=0 -e AITER_JIT_DIR=/tmp/aiter-jit-preroute-fp8 \
  --entrypoint bash "$IMAGE" -lc '
git clone -q https://github.com/ROCm/aiter.git /tmp/aiter && cd /tmp/aiter
git fetch -q origin pull/4504/head:pr pull/4498/head:bf16
test "$(git rev-parse pr)" = 3d1208d41108ba87cc571f8a692b6ec2f669348a
test "$(git rev-parse bf16)" = 0d755fe1548714f81497f37af282336187718d05
git checkout -q --detach pr
git restore --source=bf16 --worktree -- \
  aiter/ops/flydsl/kimi_k3_moe_preroute_bf16.py \
  aiter/ops/flydsl/kernels/kimi_k3_dual_projection_bf16_gfx950.py \
  aiter/ops/flydsl/kernels/kimi_k3_shared_down_bf16_gfx950.py
git submodule update -q --init --recursive
AITER_USE_SYSTEM_TRITON=1 GPU_ARCHS=gfx950 python3 setup.py develop
python3 -m pytest -q op_tests/flydsl_tests/test_kimi_k3_moe_preroute_fp8.py
'
```

## Test results

| Boundary | BF16 parent | FP8 | Speedup |
| --- | ---: | ---: | ---: |
| Dual projection + shared down | 17.6833 us | 13.8263 us | **1.2790x** |
| Dual projection + router linear | 20.3035 us | 10.6829 us | **1.9006x** |

| Output versus BF16 parent | RRMSE | Cosine |
| --- | ---: | ---: |
| Routed | 0.02616 | 0.99966 |
| Shared down | 0.04844 | 0.99884 |
| Tri shared gate/up | 0.02734 | 0.99963 |
| Tri router logits | 0 | 1.0 |

- Focused suite: 6/6 passed; strict 174.0/167.0 selection boundary and exact
  top-16 membership reproduced.
- Paired vLLM #50665 head `84a9b03c803421804eb3062de43022edce6a8992`
  on `moonshotai/Kimi-K3@9f62e4e9fffbd0a83ddd60e1c209d828994b3569`:
  full GSM8K 1,270/1,319 parent versus 1,272/1,319 candidate, zero invalid,
  paired `p=0.8450`.

## Overlap and limits

`torch.topk` does not order equal logits; the test compares the selected set
and a strict boundary, not tied ranks. Operator gates do not establish model
quality, so full-model evidence is reported separately. #4498 owns the BF16
parent; vLLM #50665 owns weight packing and dispatch.

## Tool assistance

Cursor assisted with the initial FP8 implementation. OpenAI Codex assisted
with implementation, tests, benchmarking, and drafting this description.

## Long-prefix agentic serving gate (2026-08-02)

The FP8 path was compared with the exact BF16 #4498 intermediate under the
same vLLM #50665 candidate source in the `aiperf==0.11.0` long-prefix sweep:
eight 63,240-token prefixes, 4,760 synthetic input tokens, exactly 350 output
tokens, concurrency 1/8/16/24/48, requests 5/40/80/120/240, and seed 42 on
MI355X TP8.

Both arms completed **485/485** requests with identical input hashes, exact
350-token outputs, and zero errors or endpoint faults.

| Concurrency | Output tok/s, BF16 -> FP8 | Change | Mean ITL, ms | Mean TTFT, ms | Mean E2E, ms |
|---:|---:|---:|---:|---:|---:|
| 1 | 14.27 -> 14.41 | **+1.02%** | 50.96 -> 50.13 (-1.61%) | 6,745 -> 6,785 (+0.59%) | 24,528 -> 24,281 (-1.01%) |
| 8 | 83.23 -> 84.38 | **+1.37%** | 77.70 -> 78.02 (+0.41%) | 6,170 -> 5,714 (-7.39%) | 33,286 -> 32,941 (-1.04%) |
| 16 | 145.03 -> 145.36 | +0.23% | 86.19 -> 87.01 (+0.95%) | 8,114 -> 7,632 (-5.94%) | 38,195 -> 37,999 (-0.51%) |
| 24 | 191.62 -> 191.44 | -0.09% | 95.72 -> 94.91 (-0.85%) | 9,654 -> 10,069 (+4.30%) | 43,059 -> 43,191 (+0.31%) |
| 48 | 296.46 -> 296.51 | +0.02% | 125.59 -> 125.15 (-0.36%) | 11,970 -> 12,154 (+1.53%) | 55,803 -> 55,830 (+0.05%) |

The incremental FP8 path is positive at concurrency one and eight and neutral
at higher concurrency. The exact `(1, 7168)` support predicate explains the
expected fallback once scheduler batches exceed one; this is an inference from
the source contract because no dispatch counter was collected. TTFT is
queueing-sensitive and receives no aggregate improvement claim. Each cell is
one profile trial. All four #4504 runtime files at current head `3d1208d4` are
SHA-256-identical to the measured AITER runtime.

### CODE REVIEW

File: op_tests/flydsl_tests/test_kimi_k3_moe_preroute_fp8.py [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+        )
+        .to(torch.bfloat16)
+        .float()
+    )
+    shared_ref = activated @ shared_down_dequant.t()
+
+    assert _relative_rmse(routed, routed_ref) < 0.035
+    assert _relative_rmse(gate_up, gate_up_ref) < 0.035
+    torch.testing.assert_close(tri_routed, routed, atol=0, rtol=0)
+    torch.testing.assert_close(tri_gate_up, gate_up, atol=0, rtol=0)
```
[the remark this answers is not in the document]
Comment (JohnQinAMD) [reply op_tests/flydsl_tests/test_kimi_k3_moe_preroute_fp8.py]: Agreed. Commit 7876a069 no longer requires bitwise equality between different reduction schedules. Both tri-projection outputs are now checked independently against the same dequantized reference with the existing relative-RMSE bounds. The updated focused MI355X suite passes 21/21; Ruff, Black, and git diff --check also pass.
